File: samples/typescript/tests/json-schema-kb-nexus.test.ts
import { fileURLToPath } from "node:url";
import { describe, expect, test } from "vitest";
import * as nexus from "nexus-rpc";

import { knowledgeBaseService } from "../kb/kb/services.ts";
import { BlockMapper } from "../kb/content/block/models.ts";
import type { Block } from "../kb/content/block/models.ts";
import { CategoryMapper } from "../kb/tree/category/models.ts";
import type { Category } from "../kb/tree/category/models.ts";
import { PageMapper } from "../kb/content/page/models.ts";
import type { Page } from "../kb/content/page/models.ts";
import {
  GetCategoryTreeInputMapper,
  GetPageInputMapper,
  PutBlockOutputMapper,
} from "../kb/kb/models.ts";
import type { PutBlockOutput } from "../kb/kb/models.ts";
import { loadFixture as loadFixtureFrom } from "./json-converter-helper.ts";
import { executeWorkflowWithNexus, withWorkflowEnvironment } from "./helpers.ts";

const workflowsPath = fileURLToPath(
  new URL("./workflows/json-schema-kb.ts", import.meta.url),
);
const wireFixtureDir = new URL("../../wire/json_schema/kb/", import.meta.url);

function loadFixture<T = unknown>(name: string): T {
  return loadFixtureFrom<T>(wireFixtureDir, name);
}

describe("json-schema KB generated Nexus service", () => {
  test("exposes generated service + operation definitions", () => {
    expect(knowledgeBaseService.name).toBe("example.kb.v1.KnowledgeBaseService");
    expect(knowledgeBaseService.operations.getPage.name).toBe("GetPage");
    expect(knowledgeBaseService.operations.putBlock.name).toBe("PutBlock");
    expect(knowledgeBaseService.operations.getCategoryTree.name).toBe(
      "GetCategoryTree",
    );
  });

  // Register a handler bound to the generated service definition, then run a
  // workflow that calls every operation through the SDK's Nexus client. Both
  // sides bridge the Nexus wire payloads through the generated model mappers:
  // the handler validates each request with `fromIntermediate` and projects
  // each response with `toIntermediate`, exercising the generated
  // service/operation definitions end-to-end over a real Temporal + Nexus
  // endpoint.
  test("drives every operation through a real Nexus client", async () => {
    await withWorkflowEnvironment(async (env) => {
      const calls: Array<[string, unknown]> = [];
      const handler = nexus.serviceHandler(knowledgeBaseService, {
        async getPage(_ctx, input) {
          calls.push(["GetPage", new GetPageInputMapper().fromIntermediate(input)]);
          return new PageMapper().toIntermediate(
            loadFixture<Page>("page.json"),
          ) as Page;
        },
        async putBlock(_ctx, input) {
          calls.push(["PutBlock", new BlockMapper().fromIntermediate(input)]);
          return new PutBlockOutputMapper().toIntermediate(
            loadFixture<PutBlockOutput>("put-block-output.json"),
          ) as PutBlockOutput;
        },
        async getCategoryTree(_ctx, input) {
          calls.push([
            "GetCategoryTree",
            new GetCategoryTreeInputMapper().fromIntermediate(input),
          ]);
          return new CategoryMapper().toIntermediate(
            loadFixture<Category>("category-tree.json"),
          ) as Category;
        },
      });

      const result = await executeWorkflowWithNexus<{
        blockId: string;
        categoryChildId: string | undefined;
        pageId: string;
        revision: number;
      }>(env, {
        endpoint: "knowledge-base",
        nexusServices: [handler],
        workflowType: "jsonSchemaKbCaller",
        workflowsPath,
      });

      expect(result).toEqual({
        blockId: "block-1",
        categoryChildId: "child",
        pageId: "page-1",
        revision: 7,
      });

      expect(calls.map(([operation]) => operation)).toEqual([
        "GetPage",
        "PutBlock",
        "GetCategoryTree",
      ]);
      expect((calls[0]?.[1] as { pageId: string }).pageId).toBe("page-1");
      const putBlockInput = calls[1]?.[1] as Block;
      expect(putBlockInput.blockId).toBe("block-1");
      expect(putBlockInput.style?.bold).toBe(true);
      expect((calls[2]?.[1] as { rootId: string }).rootId).toBe("root");
    });
  });
});

File: samples/java/src/test/java/jsonschema/JsonSchemaKbNexusTest.java
package jsonschema;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.protobuf.ByteString;
import io.nexusrpc.handler.OperationHandler;
import io.nexusrpc.handler.OperationImpl;
import io.nexusrpc.handler.ServiceImpl;
import io.temporal.api.common.v1.Payload;
import io.temporal.client.WorkflowOptions;
import io.temporal.common.converter.DataConverter;
import io.temporal.common.converter.DefaultDataConverter;
import io.temporal.testing.TestWorkflowEnvironment;
import io.temporal.worker.Worker;
import io.temporal.workflow.NexusOperationOptions;
import io.temporal.workflow.NexusServiceOptions;
import io.temporal.workflow.Workflow;
import io.temporal.workflow.WorkflowInterface;
import io.temporal.workflow.WorkflowMethod;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.junit.jupiter.api.Test;

import json_schema.definitions.kb.content.block.Block;
import json_schema.definitions.kb.content.page.Page;
import json_schema.definitions.kb.kb.GetCategoryTreeInput;
import json_schema.definitions.kb.kb.GetPageInput;
import json_schema.definitions.kb.kb.KnowledgeBaseService;
import json_schema.definitions.kb.kb.PutBlockOutput;
import json_schema.definitions.kb.tree.category.Category;

/**
 * Drives the generated {@link KnowledgeBaseService} Nexus service definition
 * end-to-end over a real Temporal + Nexus endpoint. Unlike {@link
 * JsonSchemaRoundTripTest} (which round-trips fixtures through the data
 * converter), the caller workflow here calls every operation through the
 * Temporal SDK's built-in Nexus service stub — there is no generated API
 * client — exercising both the generated service/operation definitions and the
 * models over the wire in both directions.
 */
final class JsonSchemaKbNexusTest {

    private static final String ENDPOINT = "knowledge-base";
    private static final String TASK_QUEUE = "kb-nexus-task-queue";
    private static final DataConverter CONVERTER = DefaultDataConverter.newDefaultInstance();

    private static <T> T fixture(String name, Class<T> type) {
        try {
            Path dir = Paths.get(System.getProperty("user.dir"), "..", "wire", "json_schema", "kb")
                    .normalize();
            byte[] data = Files.readAllBytes(dir.resolve(name));
            Payload payload = Payload.newBuilder()
                    .putMetadata("encoding", ByteString.copyFromUtf8("json/plain"))
                    .setData(ByteString.copyFrom(data))
                    .build();
            return CONVERTER.fromPayload(payload, type, type);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    /** Handler implementing the generated service interface, backed by wire fixtures. */
    @ServiceImpl(service = KnowledgeBaseService.class)
    public static final class KnowledgeBaseServiceImpl {
        final List<String> calls = Collections.synchronizedList(new ArrayList<>());

        @OperationImpl
        public OperationHandler<GetPageInput, Page> getPage() {
            return OperationHandler.sync((ctx, details, input) -> {
                calls.add("GetPage");
                assertEquals("page-1", input.getPageId());
                return fixture("page.json", Page.class);
            });
        }

        @OperationImpl
        public OperationHandler<Block, PutBlockOutput> putBlock() {
            return OperationHandler.sync((ctx, details, input) -> {
                calls.add("PutBlock");
                assertEquals("block-1", input.getBlockId());
                assertNotNull(input.getStyle());
                assertTrue(input.getStyle().getBold());
                return fixture("put-block-output.json", PutBlockOutput.class);
            });
        }

        @OperationImpl
        public OperationHandler<GetCategoryTreeInput, Category> getCategoryTree() {
            return OperationHandler.sync((ctx, details, input) -> {
                calls.add("GetCategoryTree");
                assertEquals("root", input.getRootId());
                return fixture("category-tree.json", Category.class);
            });
        }
    }

    /** Result the caller workflow returns for assertion on the client side. */
    public static final class KbResult {
        public String blockId;
        public String categoryChildId;
        public String pageId;
        public long revision;

        public KbResult() {}

        public KbResult(String blockId, String categoryChildId, String pageId, long revision) {
            this.blockId = blockId;
            this.categoryChildId = categoryChildId;
            this.pageId = pageId;
            this.revision = revision;
        }
    }

    @WorkflowInterface
    public interface KnowledgeBaseCaller {
        @WorkflowMethod
        KbResult run();
    }

    public static final class KnowledgeBaseCallerImpl implements KnowledgeBaseCaller {
        private final KnowledgeBaseService kb = Workflow.newNexusServiceStub(
                KnowledgeBaseService.class,
                NexusServiceOptions.newBuilder()
                        .setEndpoint(ENDPOINT)
                        .setOperationOptions(NexusOperationOptions.newBuilder()
                                .setScheduleToCloseTimeout(Duration.ofSeconds(10))
                                .build())
                        .build());

        @Override
        public KbResult run() {
            Page page = kb.getPage(new GetPageInput("page-1"));
            List<Block> blocks = page.getBlocks();
            if (blocks == null || blocks.isEmpty()) {
                throw new IllegalStateException("expected page block");
            }
            PutBlockOutput putBlockOutput = kb.putBlock(blocks.get(0));
            Category category = kb.getCategoryTree(new GetCategoryTreeInput("root"));
            String categoryChildId =
                    category.getChildren() != null && !category.getChildren().isEmpty()
                            ? category.getChildren().get(0).getId()
                            : null;
            return new KbResult(
                    putBlockOutput.getBlockId(),
                    categoryChildId,
                    page.getPageId(),
                    putBlockOutput.getRevision());
        }
    }

    @Test
    void kbOperationsUseRealNexusClient() {
        TestWorkflowEnvironment env = TestWorkflowEnvironment.newInstance();
        try {
            Worker worker = env.newWorker(TASK_QUEUE);
            worker.registerWorkflowImplementationTypes(KnowledgeBaseCallerImpl.class);
            KnowledgeBaseServiceImpl handler = new KnowledgeBaseServiceImpl();
            worker.registerNexusServiceImplementation(handler);
            env.createNexusEndpoint(ENDPOINT, TASK_QUEUE);
            env.start();

            KnowledgeBaseCaller workflow = env.getWorkflowClient().newWorkflowStub(
                    KnowledgeBaseCaller.class,
                    WorkflowOptions.newBuilder().setTaskQueue(TASK_QUEUE).build());
            KbResult result = workflow.run();

            assertEquals("block-1", result.blockId);
            assertEquals("child", result.categoryChildId);
            assertEquals("page-1", result.pageId);
            assertEquals(7L, result.revision);
            assertEquals(Arrays.asList("GetPage", "PutBlock", "GetCategoryTree"), handler.calls);
        } finally {
            env.close();
        }
    }
}
